File: leetcode_cpp/best_time_to_buy_and_sell_stock_ii.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        if (prices.size() <= 1) return 0;

        int profit = 0, buy_price, sell_price, i;
        bool smaller_than_future = (prices[0] < prices[1]), smaller_than_past;
        if (smaller_than_future) buy_price = prices[0];

        for (i = 1; i < prices.size() - 1; ++i) {
            smaller_than_past = !smaller_than_future;
            smaller_than_future = (prices[i] < prices[i + 1]);
            if (!smaller_than_past && !smaller_than_future) {
                sell_price = prices[i];
                profit += (sell_price - buy_price);
            }
            else if (smaller_than_past && smaller_than_future) {
                buy_price = prices[i];
            }
        }

        i = prices.size() - 1;
        smaller_than_past = !smaller_than_future;
        if (!smaller_than_past) {
            sell_price = prices[i];
            profit += (sell_price - buy_price);
        }

        return profit;
    }
};

File: leetcode_python/best_time_to_buy_and_sell_stock_ii.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if len(prices) <= 1:
            return 0

        profit = 0
        
        smaller_than_future = (prices[0] < prices[1])
        buy_price = prices[0] if smaller_than_future else None

        for ind in range(1, len(prices) - 1):
            smaller_than_past = not smaller_than_future
            smaller_than_future = (prices[ind] < prices[ind + 1])
            if not smaller_than_past and not smaller_than_future:
                sell_price = prices[ind]
                profit += (sell_price - buy_price)
                buy_price = None
            elif smaller_than_past and smaller_than_future:
                buy_price = prices[ind]

        ind = len(prices) - 1
        smaller_than_past = not smaller_than_future
        if not smaller_than_past:
            sell_price = prices[ind]
            profit += (sell_price - buy_price)

        return profit
